### PR TITLE
Feature/detecting inconsistent weaving

### PR DESCRIPTION
**Pull request by TheCelavi**
@lisachenko Bad weaving is not possible to detect in runtime.

But it is possible to detect it via command. If there is a "bad" weaving (circular references and what now) -> generated proxies will vary. 

So -> at least, we can provide a command utility that can help in debugging.

Need to do some code cleanup - but I would like you to checkout idea and see if we are on same page.

### CODE REVIEW

File: tests/Go/Functional/BaseFunctionalTest.php [hunk 1]
```diff
@@ -40,7 +43,8 @@ protected static function exec($command, $args = '')
 
         $process->run();
 
-        self::assertTrue($process->isSuccessful(), sprintf('Unable to execute "%s" command.', $command));
+        $assert = ($expectSuccess) ? 'assertTrue' : 'assertFalse';
+        self::{$assert}($process->isSuccessful(), sprintf('Unable to execute "%s" command, got output: "%s".', $command, $process->getOutput()));
```
Comment (lisachenko) [inline tests/Go/Functional/BaseFunctionalTest.php]: Better not to use dynamic static method, just write if..else construction, it will be more readable

File: tests/Go/Console/Command/DebugWeavingCommandTest.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace Go\Console\Command;
+
+use Go\Functional\BaseFunctionalTest;
+
+class DebugWeavingCommandTest extends BaseFunctionalTest
```
Comment (lisachenko) [inline tests/Go/Console/Command/DebugWeavingCommandTest.php]: I'm trying to understand this test, how it's possible to weave differently each time?

File: tests/Fixtures/project/src/Aspect/InconsistentlyWeavingAspect.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+namespace Go\Tests\TestProject\Aspect;
+
+use Go\Aop\Aspect;
+use Go\Aop\Intercept\MethodInvocation;
+use Go\Tests\TestProject\Application\InconsistentlyWeavedClass;
+use Go\Lang\Annotation as Pointcut;
+
+class InconsistentlyWeavingAspect implements Aspect
+{
+    public function __construct(InconsistentlyWeavedClass $problematicClass)
```
Comment (lisachenko) [inline tests/Fixtures/project/src/Aspect/InconsistentlyWeavingAspect.php]: Ah, yes )) Injection of intercepted class, please leave a note about this edge-case in the doc block and link to the issue on github why this happen.

File: tests/Fixtures/project/src/Aspect/InconsistentlyWeavingAspect.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    {
+    }
+
+    /**
+     * Intercepts badlyWeaved()
+     *
+     * @param MethodInvocation $invocation
+     *
+     * @Pointcut\After("execution(public Go\Tests\TestProject\Application\InconsistentlyWeavedClass->badlyWeaved(*))")
```
Comment (lisachenko) [inline tests/Fixtures/project/src/Aspect/InconsistentlyWeavingAspect.php]: Make a note here as well, that we are trying to intercept method from the class that we want to use in the current aspect.

File: src/Console/Command/DebugWeavingCommand.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+use Go\Instrument\ClassLoading\CachePathManager;
+use Go\Instrument\ClassLoading\CacheWarmer;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\NullOutput;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+/**
+ * Console command for querying an information about aspects
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Doc block should be changed according to the functionality of current command

File: src/Console/Command/DebugWeavingCommand.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->loadAspectKernel($input, $output);
+
+        $io = new SymfonyStyle($input, $output);
+
+        $io->title('Weaving debug information');
+
+        $cachePathManager = $this->aspectKernel->getContainer()->get('aspect.cache.path.manager');
+        $warmer = new CacheWarmer($this->aspectKernel, new NullOutput());
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Please, enable assignment alignment for the project in your IDE

File: src/Console/Command/DebugWeavingCommand.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+Allows to check consistency of weaving process, detects circular references and mutual dependencies between
+subjects of weaving and aspects.
+EOT
+            );
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php:46]: Idea is clever 👍

File: src/Console/Command/DebugWeavingCommand.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        $proxies = $this->getProxies($cachePathManager);
+
+        $cachePathManager->clearCacheState();
+        $warmer->warmUp();
+
+        $errors = 0;
+
+        foreach ($this->getProxies($cachePathManager) as $path => $content) {
+            if (!isset($proxies[$path])) {
+                $io->writeln(sprintf('<fg=white;bg=red;options=bold>[ERR]</>: Proxy on path "%s" is generated on second "warmup" pass.', $path));
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Maybe just switch to `$io->error()` or `$io->caution()` here? Instead of manual formatting?

File: src/Console/Command/DebugWeavingCommand.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                continue;
+            }
+
+            if (isset($proxies[$path]) && $proxies[$path] !== $content) {
+                $io->writeln(sprintf('<fg=white;bg=red;options=bold>[ERR]</>: Proxy on path "%s" is weaved differnlty on second "warmup" pass.', $path));
+                $errors++;
+                continue;
+            }
+
+            $io->writeln(sprintf('<fg=green;options=bold>[OK]</>: <comment>Proxy  on path "%s" is consistently weaved.</comment>', $path));
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Suggest is to use `$io->success()` or we can define our own formatting style. It's always cool to separate templating code from the data.

File: src/Console/Command/DebugWeavingCommand.php [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        if ($errors > 0) {
+            $io->error(sprintf('Weaving is unstable, there are %s reported error(s).', $errors));
+            return -1;
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Typically, exit codes are greater than zero, maybe switch to non-zero positive number?

File: src/Console/Command/DebugWeavingCommand.php [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        $io->success('Weaving is stabile, there are no errors reported.');
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Typo: stabile => stable.

File: src/Console/Command/DebugWeavingCommand.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    private function getProxies(CachePathManager $cachePathManager)
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Missed doc block for the method?

File: src/Console/Command/DebugWeavingCommand.php [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        $path = $cachePathManager->getCacheDir() . '/_proxies';
+        $iterator = new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($path, \FilesystemIterator::SKIP_DOTS | \FilesystemIterator::UNIX_PATHS), \RecursiveIteratorIterator::CHILD_FIRST);
+        $proxies = [];
```
Comment (lisachenko) [inline src/Console/Command/DebugWeavingCommand.php]: Assignment alignment 😄

**Review by lisachenko — CHANGES_REQUESTED** [review]

**Comment by TheCelavi** [comment]
I did setup this project for assignment alignment, I just forgot to do reformatting. 

I want to cleanup more these functional tests and to make their usage quite easy for us, we are really getting benefits from them - like, it is really, really easy to construct edge case, to test it, and to make a patch for it.

Ok, cool! Glad you like the idea, I think that, beside documentation, it is quite important that it is possible to debug these kind of issues.

It solves this one: https://github.com/goaop/goaop-symfony-bundle/issues/15, per example...

File: tests/Go/Functional/BaseFunctionalTest.php [hunk 14]
```diff
@@ -1,47 +1,167 @@
...
+     * @param bool $expectSuccess Should command be executed successfully
+     * @param null|int $expectedExitCode If provided, exit code will be asserted.
+     *
+     * @return string Console output.
+     */
+    protected function execute($command, $args = null, $expectSuccess = true, $expectedExitCode = null)
     {
-        $commandStatement = sprintf('php %s %s %s %s',
-            self::$consolePath,
+        $commandStatement = sprintf('GO_AOP_CONFIGURATION=%s php %s %s %s %s',
```
[the remark this answers is not in the document]
Comment (TheCelavi) [reply tests/Go/Functional/BaseFunctionalTest.php:113]: Not applicable.

**Comment by TheCelavi** [comment]
@lisachenko Ok, code cleanup is done, I have optimized all functional tests as well. 

Almost everything is implemented from: https://github.com/goaop/framework/issues/340, there is:

- `assertClassIsWeaved($class)`
- `assertClassIsNotWeaved($class)`

I am planning to implement

- `assertClassHasStaticMethodJoinPoint($class, $methodName, $joinPoint, $index = null)`
- `assertClassHasMethodJoinPoint($class, $methodName, $joinPoint, $index = null)`

and some more methods, as needed. But I will do that in different PR - this can be merged.

**Comment by lisachenko** [comment]
@TheCelavi  Looks good now!

PS. Just noticed that the verb "weave" is irregular, so correct form for "class is weaved" is "class is woven".

There are many places in the source code with "weaved" :(. Should we adjust your method naming to the correct one right now? Or in separate PR?
  - assertClassIsWoven($className)
  - assertClassIsNotWoven($className)

**Review by lisachenko — APPROVED** [review]

**Comment by TheCelavi** [comment]
Noup, it is easy-peasy, fixed, thanks!

**Comment by lisachenko** [comment]
HHVM has segfaulted on Travis-CI :) Time to remove it from matrix... Only delays check.

**Comment by lisachenko** [comment]
Squashed and merged, thanks!
